scripts/utils.js

```diff
@@ -85,6 +86,87 @@ async function spawnCommand(command, args) {
   })
 }
 
+/**
+ * Helper to run executables. This has been introduced to work around Shell injections[0] while
+ * keeping a good developer experience. The template string is split on white spaces and passed to
+ * a child_process method that don't use a shell.
+ *
+ * Prefer this helper over using other child_process methods. Avoid using `child_process.exec`,
+ * `child_process.execSync` or the `shell` option of other `child_process` functions.
+ *
+ * Use this as a tagged string template. Example:
+ *
+ * > command`git ls-files ${directory}`.run()
+ *
+ * [0]: https://matklad.github.io/2021/07/30/shell-injection.html
+ */
+function command(...templateArguments) {
+  const [commandName, ...commandArguments] = parseCommandArguments(...templateArguments)
+
+  let input = ''
+  let env
+
+  return {
+    withInput(newInput) {
+      input = newInput
+      return this
+    },
+
+    withEnvironment(newEnv) {
+      env = newEnv
+      return this
+    },
+
+    run() {
+      const commandResult = childProcess.spawnSync(commandName, commandArguments, {
+        input,
+        env,
+        encoding: 'utf-8',
+      })
+
+      if (commandResult.status !== 0) {
+        const formattedCommand = `${commandName} ${commandArguments.join(' ')}`
+        const formattedStderr = `----\n${commandResult.stderr || ''}\n----`
+        const exitCause = commandResult.signal ? `signal ${commandResult.signal}` : `status ${commandResult.status}`
+        const error = new Error(`Command failed with exit ${exitCause}: ${formattedCommand}\n${formattedStderr}`)
+        throw error
+      }
+
+      if (commandResult.stderr) {
+        printError(commandResult.stderr)
+      }
+
+      return commandResult.stdout
+    },
+  }
+}
+
+function parseCommandArguments(templateStrings, ...templateVariables) {
+  let parsedArguments = []
+  for (let i = 0; i < templateStrings.length; i += 1) {
+    if (i > 0) {
+      // Interleave variables with template strings
+      if (!parsedArguments.length || templateStrings[i - 1].match(/\s$/)) {
+        // If the latest string ends with a space, consider the variable as a separate argument
+        parsedArguments.push(templateVariables[i - 1])
+      } else {
+        // Else, append the variable to the latest argument
+        parsedArguments[parsedArguments.length - 1] += templateVariables[i - 1]
+      }
+    }
+
+    const words = templateStrings[i].split(/\s+/).filter((word) => word)
+
+    if (parsedArguments.length && words.length && !templateStrings[i].match(/^\s/)) {
+      // If the string does not start with a space, append it to the latest argument
+      parsedArguments[parsedArguments.length - 1] += words.shift()
+    }
+
+    parsedArguments.push(...words)
+  }
+  return parsedArguments
+}
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏it could be nice to express what we are trying to do or to avoid with this function.